client/src/game/reducers/cardDeductionReducer.ts

```diff
@@ -0,0 +1,131 @@
+  }
+};
+
+export default cardDeductionReducer;
+
+const makeDeductions = (
+  hands: ReadonlyArray<readonly number[]>,
+  deck: readonly CardState[],
+  variant: Variant,
+  metadata: GameMetadata,
+) => {
+  const newDeck: CardState[] = deck.concat([]);
+  const cardCountMap = getCardCountMap(variant);
+
+  // if our variant shows played/discarded cards, then they're known by everyone
+  newDeck.filter((card) => !hands.some((hand) => hand.includes(card.order)))
+    .filter((card) => card.suitIndex !== null && card.rank !== null)
+    .forEach((card) => {
+      cardCountMap[card.suitIndex!][card.rank!] -= 1;
+    });
+
+  // We need to calculate our own hand first because those possibilities will be needed for
+  // pretending like we know what the other players see in our hand.
+  calculateHandPossibilities(hands[metadata.ourPlayerIndex], hands, newDeck, cardCountMap);
+  hands.filter((hand) => hand !== hands[metadata.ourPlayerIndex])
+    .forEach((hand) => { calculateHandPossibilities(hand, hands, newDeck, cardCountMap); });
+  return newDeck;
+};
+
+const calculateHandPossibilities = (
+  hand: readonly number[],
+  hands: ReadonlyArray<readonly number[]>,
+  deck: CardState[],
+  cardCountMap: readonly number[][],
+) => {
+  const remainingCardCountMap = Array.from(
+    cardCountMap,
+    (arr) => Array.from(arr),
+  );
+  const remainingCards: number[] = [];
+  const handCards: number[] = [];
+  hands.forEach((otherHand) => otherHand.forEach((o) => {
+    const deckCard = deck[o];
+    handCards.push(o);
+    // If card is not fully known then we need to run through the permutations.
+    // Otherwise, it's a card in another hand and we know we don't have that copy.
+    if (deckCard.suitIndex === null || deckCard.rank === null || otherHand === hand) {
+      remainingCards.push(o);
+    } else {
+      remainingCardCountMap[deckCard.suitIndex][deckCard.rank] -= 1;
+    }
+  }));
+  hand.forEach((o) => {
+    const card = deck[o];
+    const otherRemainingCards = remainingCards.filter((ro) => ro !== o).map((ro) => deck[ro]);
+    deck[o] = {
+      ...card,
+      possibleCardsFromDeduction: card.possibleCardsFromDeduction.filter(
+        (possibility) => possibilityValid(possibility, otherRemainingCards, remainingCardCountMap),
+      ),
+    };
+  });
+};
+
+const possibilityValid = (
+  [suit, rank]: readonly [number, number],
+  remainingCards: CardState[],
+  cardCountMap: readonly number[][],
+) => {
+  // Avoiding duplicating the map for performance, so trying to undo the mutation as we exit
+  cardCountMap[suit][rank] -= 1;
+  if (cardCountMap[suit][rank] >= 0 && remainingCardsValid(remainingCards, cardCountMap)) {
+    cardCountMap[suit][rank] += 1;
+    return true;
+  }
+  cardCountMap[suit][rank] += 1;
+  return false;
+};
+
+const remainingCardsValid = (
+  remainingCards: readonly CardState[],
+  cardCountMap: readonly number[][],
+) => remainingCards.length === 0 || remainingCards[0].possibleCardsFromDeduction.some(
```

Comment:
break this into multiple lines

Reply:
Also why is this passing an array of card states around? card states are very big, so this is basically making multiple copies of the deck?

Reply:
(on every action)

Reply:
I don't disagree that I only need possibleCardsFromDeduction for this method, so I'll fix that.  I do want to understand the comment about copies better.  I thought javascript was pass-by-value of the reference and therefore wasn't copying the whole CardState, just copying the reference value of the CardState (or in this case, a copy of the reference to the array).

If I add a `remainingCards.pop()` call to the end of the `possibilityValid` method, then tests fail.  If it were copying the entire array, as opposed to copying the reference value, then I would think it would have no impact (since it's at the end of the method)?

```
const possibilityValid = (
  [suit, rank]: readonly [number, number],
  remainingCards: CardState[],
  cardCountMap: readonly number[][],
) => {
  // Avoiding duplicating the map for performance, so trying to undo the mutation as we exit
  cardCountMap[suit][rank] -= 1;
  if (cardCountMap[suit][rank] >= 0 && remainingCardsValid(remainingCards, cardCountMap)) {
    cardCountMap[suit][rank] += 1;
    return true;
  }
  // We're no longer using remainingCards inside this method after the pop call
  remainingCards.pop();
  cardCountMap[suit][rank] += 1;
  return false;
};
```

Reply:
I think you're right Smandon. Passing arrays around should be as performant as passing integers.

Reply:
note that it depends whether you are passing a vanilla array or a redux draft array.
if you are using redux anything arrays, then i think its worse